File: Python/bratchoku.py
import time
from board_helper import boardHelper


class Bratchoku:
    def __init__(self):
        self.simulated_moves = 0
        self.board_helper = boardHelper()

    def play(self, black_bitboard: int, white_bitboard: int, depth: int):
        self.simulated_moves = 0
        start = time.time()
        
        black_to_play = black_bitboard.bit_count() == white_bitboard.bit_count()
        
        moves = self.board_helper.get_sorted_moves(black_bitboard, white_bitboard, black_to_play)
        
        best_move = -1
        alpha = -float('inf')
        beta = float('inf')
        if black_to_play:
            best_val = -float('inf')
            for m in moves:
                black_bitboard |= (1 << m)
                val = self.minimax(black_bitboard, white_bitboard, depth - 1, False, alpha, beta)
                self.simulated_moves += 1
                black_bitboard &= ~(1 << m)
                
                if val > best_val:
                    best_val = val
                    best_move = m
                alpha = max(alpha, best_val)
        else:
            best_val = float('inf')
            for m in moves:
                white_bitboard |= (1 << m)
                val = self.minimax(black_bitboard, white_bitboard, depth - 1, True, alpha, beta)
                self.simulated_moves += 1
                white_bitboard &= ~(1 << m)
                
                if val < best_val:
                    best_val = val
                    best_move = m
                beta = min(beta, best_val)

        move_str = self.board_helper.index_to_notation(best_move)
        
        return move_str, time.time() - start, self.simulated_moves
    
    def minimax(self, black_bb, white_bb, depth, black_to_play, alpha, beta):

        if not black_to_play:
            if self.board_helper.is_winning_position(black_bb):
                return 1000000 + depth
        else:
            if self.board_helper.is_winning_position(white_bb):
                return -1000000 - depth
            
        combined = black_bb | white_bb
        if depth == 0 or combined == 0xFFFFFFFFFFFFFFFF: # board plein
            return self.board_helper.evaluate_board(black_bb, white_bb)
        
        moves = self.board_helper.get_sorted_moves(black_bb, white_bb, black_to_play)

        if black_to_play:
            max_eval = -float('inf')
            for move_index in moves:
                move_mask = 1 << move_index
                
                black_bb |= move_mask
                eval_score = self.minimax(black_bb, white_bb, depth - 1, False, alpha, beta)
                self.simulated_moves += 1
                black_bb &= ~move_mask
                
                max_eval = max(max_eval, eval_score) # type: ignore
                alpha = max(alpha, eval_score) # type: ignore
                if beta <= alpha:
                    break
            return max_eval

        else:
            min_eval = float('inf')
            for move_index in moves:
                move_mask = 1 << move_index
                
                white_bb |= move_mask
                eval_score = self.minimax(black_bb, white_bb, depth - 1, True, alpha, beta)
                self.simulated_moves += 1
                white_bb &= ~move_mask
                
                min_eval = min(min_eval, eval_score)
                beta = min(beta, eval_score)
                if beta <= alpha:
                    break
            return min_eval

File: C#/GomokuAI/Bratchoku.cs
using System.Diagnostics;
using System.Numerics;

namespace GomokuAI
{
    public class Bratchoku
    {
        public long SimulatedMoves { get; private set; }
        private readonly BoardHelper _boardHelper;

        public Bratchoku()
        {
            SimulatedMoves = 0;
            _boardHelper = new BoardHelper();
        }

        public (string move, double time, long nodes) Play(ulong blackBitboard, ulong whiteBitboard, int depth)
        {
            SimulatedMoves = 0;
            var watch = Stopwatch.StartNew();

            // bit_count() -> PopCount
            bool blackToPlay = BitOperations.PopCount(blackBitboard) == BitOperations.PopCount(whiteBitboard);

            int[] moves = (int[])_boardHelper.GetSortedMoves(blackBitboard, whiteBitboard, blackToPlay);

            int bestMove = -1;
            double alpha = double.NegativeInfinity;
            double beta = double.PositiveInfinity;

            if (blackToPlay)
            {
                double bestVal = double.NegativeInfinity;
                foreach (int m in moves)
                {
                    blackBitboard |= (1UL << m);
                    double val = Minimax(blackBitboard, whiteBitboard, depth - 1, false, alpha, beta);
                    SimulatedMoves++;
                    blackBitboard &= ~(1UL << m);

                    if (val > bestVal)
                    {
                        bestVal = val;
                        bestMove = m;
                    }
                    alpha = Math.Max(alpha, bestVal);
                }
            }
            else
            {
                double bestVal = double.PositiveInfinity;
                foreach (int m in moves)
                {
                    whiteBitboard |= (1UL << m);
                    double val = Minimax(blackBitboard, whiteBitboard, depth - 1, true, alpha, beta);
                    SimulatedMoves++;
                    whiteBitboard &= ~(1UL << m);

                    if (val < bestVal)
                    {
                        bestVal = val;
                        bestMove = m;
                    }
                    beta = Math.Min(beta, bestVal);
                }
            }

            watch.Stop();
            string moveStr = _boardHelper.IndexToNotation(bestMove);

            return (moveStr, watch.Elapsed.TotalSeconds, SimulatedMoves);
        }

        private double Minimax(ulong blackBB, ulong whiteBB, int depth, bool blackToPlay, double alpha, double beta)
        {
            if (!blackToPlay)
            {
                if (_boardHelper.IsWinningPosition(blackBB)) return 1000000 + depth;
            }
            else
            {
                if (_boardHelper.IsWinningPosition(whiteBB)) return -1000000 - depth;
            }

            ulong combined = blackBB | whiteBB;
            if (depth == 0 || combined == 0xFFFFFFFFFFFFFFFF)
            {
                return _boardHelper.EvaluateBoard(blackBB, whiteBB);
            }

            int[] moves = (int[])_boardHelper.GetSortedMoves(blackBB, whiteBB, blackToPlay);

            if (blackToPlay)
            {
                double maxEval = double.NegativeInfinity;
                foreach (int moveIndex in moves)
                {
                    ulong moveMask = 1UL << moveIndex;
                    blackBB |= moveMask;
                    double evalScore = Minimax(blackBB, whiteBB, depth - 1, false, alpha, beta);
                    SimulatedMoves++;
                    blackBB &= ~moveMask;

                    maxEval = Math.Max(maxEval, evalScore);
                    alpha = Math.Max(alpha, evalScore);
                    if (beta <= alpha) break;
                }
                return maxEval;
            }
            else
            {
                double minEval = double.PositiveInfinity;
                foreach (int moveIndex in moves)
                {
                    ulong moveMask = 1UL << moveIndex;
                    whiteBB |= moveMask;
                    double evalScore = Minimax(blackBB, whiteBB, depth - 1, true, alpha, beta);
                    SimulatedMoves++;
                    whiteBB &= ~moveMask;

                    minEval = Math.Min(minEval, evalScore);
                    beta = Math.Min(beta, evalScore);
                    if (beta <= alpha) break;
                }
                return minEval;
            }
        }
    }
}